### PR TITLE
Add `on conflict` clause to query pipeline

### LINKED ISSUE #36283: NPE for `on-conflict-clause`
**Description:**
$title.

**Steps to reproduce:**
```ballerina
import ballerina/io;

type Employee record {|
    readonly int id;
    string firstName;
    string lastName;
    int salary;
|};

public function main() {
    Employee[] employees = [
        {id: 1, firstName: "John", lastName: "Smith", salary: 1000},
        {id: 1, firstName: "John", lastName: "Smith", salary: 1000},
        {id: 2, firstName: "Fred", lastName: "Bloggs", salary: 2000}
    ];

    var highPaidEmployees = table key(id) from var e in employees 
        where e.salary >= 1000
        select e
        on conflict error("abc " + e.toString()); // Note. Works properly when remove `e.toString()`.
    
    io:println(highPaidEmployees);
}
```

Error
```
[2022-05-23 18:11:16,642] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:169)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:452)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$4(BIROptimizer.java:131)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:131)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:410)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:108)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:108)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:88)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:84)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:272)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:447)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:381)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:165)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:134)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:112)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:169)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:111)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:122)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:205)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [.:(1:1,1:1)] Compilation failed due to an unhandled exception
```
**Affected Versions:**
 2201.1.0 (Swan Lake)

### LINKED ISSUE #38991: [Bug]: Using a frame variable in `on conflict` clause gives `NPE` in
### Description

Consider the following sample where it used frame variable `name` when constructing on conflict error.

```bal
type Student record {|
    readonly int id;
    string name;
    int score;
|};

public function main() {
    table<Student> students = table [
            {id: 1, name: "John", score: 100},
            {id: 2, name: "Jane", score: 150},
            {id: 1, name: "John", score: 200}
        ];

    // Result of the following will be an error since key `John` is be duplicated
    map<int>|error studentScores = map from var {name, score} in students
        select [name, score]
        on conflict error("Key Conflict", detail = "duplicate key " + name);
}
```
<details>
  <summary>Stack trace.(Click to expand)</summary>

```
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in
https://github.com/ballerina-platform/ballerina-lang/issues with the
log you get below and your sample code.

We thank you for helping make us better.

[2022-12-08 14:32:40,368] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:159)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:464)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$3(BIROptimizer.java:121)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:121)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:422)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:107)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:107)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:86)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:83)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:287)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:451)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:385)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:170)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:139)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:117)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:170)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:116)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:165)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:213)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [.:(1:1,1:1)] Compilation failed due to an unhandled exception
error: compilation contains errors
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #41525: [Bug]: Compiler bug when a variable defined in a query expression is used in `on conflict`
### Description

This bug is caused when a variable defined in a query expression is used in an `on conflict` clause.

### Steps to Reproduce

Run the following code:
```
string[] foo = ["test", "another test", "final test", "test"];

public function main() returns error? {
    // checking for duplicates
    map<boolean> _ = check map from string f in foo
                        select [f, true]
                        on conflict error(string`Found duplicate ${f}`);
}
```

### Affected Version(s)

Ballerina 2201.7.2 (Swan Lake Update 7)
Language specification 2023R1
Update Tool 1.4.0

### OS, DB, other environment details and versions

MacOS Ventura 13.5.2

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #41525 
Fixes #38991
Fixes #36283

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 1]
```diff
@@ -924,6 +936,28 @@ BLangVariableReference addSelectFunction(BLangBlockStmt blockStmt, BLangSelectCl
...
+     */
+    BLangVariableReference addOnConflictFunction(BLangBlockStmt blockStmt, BLangOnConflictClause onConflictClause,
+                                                 List<BLangStatement> stmtsToBePropagated) {
+        Location pos = onConflictClause.pos;
+        BLangLambdaFunction lambda = createPassthroughLambda(pos);
+        BLangBlockFunctionBody body = (BLangBlockFunctionBody) lambda.function.body;
+        body.stmts.addAll(0, stmtsToBePropagated);
+        BVarSymbol oldFrameSymbol = lambda.function.requiredParams.get(0).symbol;
+        BLangSimpleVarRef frame = ASTBuilderUtil.createVariableRef(pos, oldFrameSymbol);
+        // $frame#[$error$] = on-conflict-expr;
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:955]: This is added for understanding purpose.

**Comment by KavinduZoysa** [comment]
@LakshanWeerasinghe, please fix failing tests.

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 2]
```diff
@@ -73,6 +73,11 @@ function createSelectFunction(function(_Frame _frame) returns _Frame|error? sele
     return new _SelectFunction(selectFunc);
 }
 
+function createOnConflictFunction(function(_Frame _frame) returns _Frame|error? onConflictFunc)
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/helpers.bal:76]: This function can convert into a arrow function

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 3]
```diff
@@ -93,6 +98,10 @@ function getStreamFromPipeline(_StreamPipeline pipeline) returns stream<Type, Co
     return pipeline.getStream();
 }
 
+function getStreamForOnConflictFromPipeline(_StreamPipeline pipeline) returns stream<Type, CompletionType> {
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/helpers.bal:101]: This function can convert into a arrow function

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 4]
```diff
@@ -168,52 +178,85 @@ function addToTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, err
...
-                return err;
-            }
             tbl.put(<map<Type>> checkpanic v.value);
         }
         v = strm.next();
     }
+    if v is error {
+        return v;
+    }
+    return tbl;
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/helpers.bal:199]: We can use a ternary expression for this

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 5]
```diff
@@ -168,52 +178,85 @@ function addToTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, err
...
+        table<map<Type>> tempTable = check createTableForOnConflict(strm, tbl2);
+        return createImmutableTable(tbl, tempTable.toArray());
+    }
+    return createTableForOnConflict(strm, tbl);
+}
+
+function createTableForOnConflict(stream<Type, CompletionType> strm, table<map<Type>> tbl) 
+    returns table<map<Type>>|error {
+    record {| Type value; |}|CompletionType v = strm.next();
+    while (v is record {| Type value; |}) {
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/helpers.bal:221]:
```suggestion
    while v is record {| Type value; |} {
```

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 6]
```diff
@@ -168,52 +178,85 @@ function addToTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, err
...
+        record {|Type v; error? err;|}|error value = trap (<record {|Type v; error? err;|}> checkpanic v.value);
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/helpers.bal:222]: We don't use checkpanic as practrics?. Is this a special case?

File: langlib/lang.query/src/main/ballerina/types.bal [hunk 7]
```diff
@@ -115,6 +115,12 @@ class _StreamPipeline {
         var strm = internal:construct(self.constraintTd, self.completionTd, itrObj);
         return strm;
     }
+    
+    public function getStreamForOnConflict() returns stream<Type, CompletionType> {
+        OnConflictIterHelper itrObj = new (self, self.constraintTd);
+        var strm = internal:construct(self.constraintTd, self.completionTd, itrObj);
+        return strm;
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/types.bal:122]:
```suggestion
        return  internal:construct(self.constraintTd, self.completionTd, itrObj);
```

File: langlib/lang.query/src/main/ballerina/types.bal [hunk 8]
```diff
@@ -931,6 +971,29 @@ class IterHelper {
...
+        _StreamPipeline p = self.pipeline;
+        _Frame|error? f = p.next();
+        if (f is _Frame) {
+            Type v = <Type>f["$value$"];
+            error? err = <error?>f["$error$"];
+            record {|Type v; error? err;|} value = {v, err};
+            return internal:setNarrowType(self.outputType, {value: value});
+        } else {
+            return f;
+        }
```
Comment (SasinduDilshara) [inline langlib/lang.query/src/main/ballerina/types.bal:993]:
```suggestion
       return f;
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-with-query-construct-type.bal [hunk 9]
```diff
@@ -1827,6 +1869,40 @@ function testInnerQueryConstructedWithCEP() {
...
+
+function getQueryResult(error onConflictError, Customer[] customerList) returns CustomerTable|error {
+    return <CustomerTable|error> table key(id, name) from var customer in customerList
+    select {
+            id: customer.id,
+            name: customer.name,
+            noOfItems: customer.noOfItems
+    }
+    on conflict onConflictError;
+}
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-with-query-construct-type.bal:1905]: Shall we add few tests with `Query actions` as well.

File: langlib/lang.query/src/main/ballerina/helpers.bal (hunk 2 above)
Comment (LakshanWeerasinghe) [reply langlib/lang.query/src/main/ballerina/helpers.bal:76]: I have followed the existing coding pattern used in this file. @MaryamZi wdyt? Should I update it? Or should I send a separate PR updating all the places?

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-with-query-construct-type.bal (hunk 9 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-with-query-construct-type.bal:1905]: No need to add tests for query actions. OnConflict clause applies only when creating tables and maps.

File: langlib/lang.query/src/main/ballerina/helpers.bal (hunk 6 above)
Comment (LakshanWeerasinghe) [reply langlib/lang.query/src/main/ballerina/helpers.bal:222]: This the pattern we have followed so far. We may have to send another PR refactoring these.

File: langlib/lang.query/src/main/ballerina/helpers.bal (hunk 2 above)
Comment (MaryamZi) [reply langlib/lang.query/src/main/ballerina/helpers.bal:76]: We can follow best practices for new code (especially given that these are internal functions) and fix/update the older code separately in a separate PR. I wouldn't consider changing old code with this particular change as a high-priority task though.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 10]
```diff
@@ -621,7 +621,7 @@ private void createFunctionMap(BLangFunction funcNode, SymbolEnv funcEnv) {
         if (funcNode.mapSymbolUpdated) {
             return;
         }
-        BLangRecordLiteral emptyRecord = ASTBuilderUtil.createEmptyRecordLiteral(funcNode.pos, symTable.mapType);
+        BLangRecordLiteral emptyRecord = ASTBuilderUtil.createEmptyRecordLiteral(funcNode.pos, symTable.mapAllType);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java:624]: @chiranSachintha, don't we have function parameter tests for this? Let's create an issue for this and add some tests.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 11]
```diff
@@ -296,23 +301,27 @@ BLangStatementExpression desugar(BLangQueryExpr queryExpr, SymbolEnv env,
...
             BLangVariableReference tableRef = addTableConstructor(queryExpr, queryBlock);
-            result = getStreamFunctionVariableRef(queryBlock,
-                    QUERY_ADD_TO_TABLE_FUNCTION, Lists.of(streamRef, tableRef, onConflictExpr, isReadonly), pos);
+            if (onConflictExpr == null) {
+                result = getStreamFunctionVariableRef(queryBlock,
+                        QUERY_ADD_TO_TABLE_FUNCTION, Lists.of(streamRef, tableRef, isReadonly), pos);
+            } else {
+                result = getStreamFunctionVariableRef(queryBlock,
+                        QUERY_ADD_TO_TABLE_FOR_ON_CONFLICT_FUNCTION, Lists.of(streamRef, tableRef, isReadonly), pos);
+            }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:311]: Only the name changes between the two, right? I would use a ternary expression with `onConflictExpr == null` to get the function's Name assigned to a variable and use that in one usage of `getStreamFunctionVariableRef`.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 12]
```diff
@@ -925,6 +937,28 @@ BLangVariableReference addSelectFunction(BLangBlockStmt blockStmt, BLangSelectCl
...
     }
 
+    /**
+     * Desugar onConflictClause to below and return a reference to created onConflict _StreamFunction.
+     * _StreamFunction onConflictFunc = createOnConflictFunction
+     * @param blockStmt
+     * @param onConflictClause
+     * @param stmtsToBePropagated
+     * @return
+     */
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:947]: Improve or remove docs.

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 13]
```diff
@@ -168,52 +178,85 @@ function addToTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, err
...
+    if isReadOnly {
+        // TODO: Properly fix readonly scenario - Issue lang/#36721
+        // In this case tbl will be an immutable table. Therefore, we will create a new mutable table. Next, we will
+        // pass the newly created table into createTableWithKeySpecifier() to add the key specifier details from the
+        // original table variable (tbl). Then the newly created table variable will be populated using createTable()
+        // and make it immutable with createImmutableTable().
+        table<map<Type>> tempTbl = table [];
+        table<map<Type>> tbl2 = createTableWithKeySpecifier(tbl, typeof(tempTbl));
+        table<map<Type>> tempTable = check createTableForOnConflict(strm, tbl2);
+        return createImmutableTable(tbl, tempTable.toArray());
```
Comment (MaryamZi) [inline langlib/lang.query/src/main/ballerina/helpers.bal:213]: Can we improve the names of the variables?

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 14]
```diff
@@ -227,6 +270,44 @@ function createMap(stream<Type, CompletionType> strm, map<Type> mp, error? err)
...
+        createImmutableValue(check createMapForOnConflict(strm, mp2));
+        return mp2;
+    }
+    return createMapForOnConflict(strm, mp);
+}
+
+function createMapForOnConflict(stream<Type, CompletionType> strm, map<Type> mp) returns map<Type>|error {
+    record {| Type value; |}|CompletionType v = strm.next();
+    while v is record {| Type value; |} {
+        record {|Type v; error? err;|}|error value = trap (<record {|Type v; error? err;|}> checkpanic v.value);
```
Comment (MaryamZi) [inline langlib/lang.query/src/main/ballerina/helpers.bal:289]: Can't we use ensureType instead?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java (hunk 12 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java]: updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41733/commits/e2c20f614268bbaea5d3752dd3ca5bb1e52bcaba)

File: langlib/lang.query/src/main/ballerina/helpers.bal (hunk 13 above)
Comment (LakshanWeerasinghe) [reply langlib/lang.query/src/main/ballerina/helpers.bal]: updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41733/commits/e2c20f614268bbaea5d3752dd3ca5bb1e52bcaba)

**Comment by SasinduDilshara** [comment]
Shall we add a one test for nested queries?

**Comment by LakshanWeerasinghe** [comment]
> Shall we add a one test for nested queries?

added in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41733/commits/a7e115e751fda48cd0938625970ff4c1cd011ace).

File: langlib/lang.query/src/main/ballerina/helpers.bal [hunk 15]
```diff
@@ -168,52 +175,83 @@ function addToTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, err
...
+        }
+        map<Type> tblValue = check value.v.ensureType();
+        error? e = trap tbl.add(tblValue);
+        error? err = value.err;
+        if e is error && err is error {
+            return err;
+        }
+        if e is error && err is () {
+            tbl.put(<map<Type>> checkpanic value.v);
+        }
```
Comment (KavinduZoysa) [inline langlib/lang.query/src/main/ballerina/helpers.bal:231]: I think it will be more readable if we can use `hasKey`.

**Comment by KavinduZoysa** [comment]
```ballerina
import ballerina/io;

type Student record {|
    readonly int id;
    string firstName;
    string lastName;
    int score;
|};

public function main() returns error? {
    table<Student> key(id) students = table [
            {id: 1, firstName: "John", lastName: "Smith", score: 100},
            {id: 2, firstName: "John", lastName: "Smith", score: 150},
            {id: 9, firstName: "Cassie", lastName: "Smith", score: 250}
        ];

    var studentScores = check map from var {firstName, score} in students
                        group by firstName
                        select [firstName, [score]]
                        on conflict error(string `${[score].length()}`);
    io:println(studentScores);
}
```

Shall we add a test case like this? Here sequence variable is used within on-clause. There is no use in the provided code, but it is better to add.

File: langlib/lang.query/src/main/ballerina/helpers.bal (hunk 15 above)
Comment (LakshanWeerasinghe) [reply langlib/lang.query/src/main/ballerina/helpers.bal:231]: We cannot use `hasKey` because we don't know the key.

**Review by KavinduZoysa — APPROVED** [review]

**Review by SasinduDilshara — APPROVED** [review]

**Pull request merged**